File: solutions/0779. K-th Symbol in Grammar/0779.cpp
class Solution {
 public:
  int kthGrammar(int N, int K) {
    if (N == 1)
      return 0;
    if (K & 1)
      return kthGrammar(N - 1, (K + 1) / 2) != 0;  // left node
    return kthGrammar(N - 1, K / 2) == 0;          // right node
  }
};

File: solutions/0779. K-th Symbol in Grammar/0779.java
class Solution {
  public int kthGrammar(int n, int k) {
    if (n == 1)
      return 0;
    if (k % 2 == 1)
      return kthGrammar(n - 1, (k + 1) / 2) == 0 ? 0 : 1; // the left node
    return kthGrammar(n - 1, k / 2) == 0 ? 1 : 0;         // the right node
  }
}
